In [2]:
import psycopg2
import pandas as pd

In [4]:

query_3_1 = f'''
    SELECT  COUNT(name)
    FROM public.vacancies'''

In [5]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [6]:
query_3_2=f'''
    SELECT  COUNT(name) 
    FROM public.employers'''

In [7]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

In [8]:
query_3_3 = f'''
    SELECT  COUNT(name) 
    FROM public.areas'''

In [9]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [10]:
query_3_4=f'''
    SELECT COUNT(name)
    from public.industries'''

In [11]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

In [12]:
query_4_1 = f'''
    SELECT
        a.name AS area,
        COUNT(v.name) AS cnt
    FROM public.vacancies v
        JOIN public.areas a ON a.id = v.area_id
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 5'''

In [13]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

In [14]:
query_4_2=f'''
    SELECT 
        COUNT(name)
    FROM public.vacancies
    WHERE salary_from IS NOT NULL or salary_to IS NOT NULL'''

In [15]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

In [16]:
query_4_3 = f'''
    SELECT
        ROUND(AVG(salary_from)),
        ROUND(AVG(salary_to))
    FROM public.vacancies'''

In [17]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

In [18]:
query_4_4=f'''
    SELECT
        schedule,
        employment,
        COUNT(name)
    FROM public.vacancies
    GROUP BY 1, 2
    ORDER BY 3 DESC'''

In [19]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

In [20]:
query_4_5 = f'''
    SELECT       
        experience,
        COUNT(name) as vacancies
    FROM public.vacancies
    GROUP BY 1
    ORDER BY 2 '''

In [21]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancies
0         Более 6 лет       1337
1           Нет опыта       7197
2       От 3 до 6 лет      14511
3  От 1 года до 3 лет      26152

In [22]:
query_5_1=f'''
    SELECT
        e.name,
        COUNT(v.name) as vacancies
    FROM public.vacancies v
        JOIN public.employers e ON e.id = v.employer_id
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 5'''
        

In [23]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  vacancies
0         Яндекс       1933
1     Ростелеком        491
2       Тинькофф        444
3           СБЕР        428
4  Газпром нефть        331

In [24]:
query_5_2=f'''
    SELECT
        a.name,
        COUNT(e.name) as employers,
        COUNT(v.name) as vacancies
    FROM public.areas a
        LEFT JOIN public.employers e ON a.id = e.area
        LEFT JOIN public.vacancies v ON a.id = v.area_id
    WHERE v.area_id IS NULL
    GROUP BY 1
    ORDER BY 3, 2 DESC
    LIMIT 5'''
        
        

In [25]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers  vacancies
0              Россия        410          0
1           Казахстан        207          0
2  Московская область         75          0
3  Краснодарский край         19          0
4            Беларусь         18          0

In [26]:
query_5_3=f'''
    SELECT 
        e.name AS employers,
        COUNT(DISTINCT v.area_id) as areas
    FROM public.employers e
        JOIN public.vacancies v ON e.id = v.employer_id
    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 5'''

In [27]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employers  areas
0                  Яндекс    181
1              Ростелеком    152
2              Спецремонт    116
3  Поляков Денис Иванович     88
4               ООО ЕФИН      71

In [28]:
query_5_4=f'''
    SELECT
        COUNT(e.name) AS employers
    FROM public.employers e
        LEFT JOIN public.employers_industries i ON e.id = i.employer_id
    WHERE i.industry_id IS NULL'''

In [29]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers
0       8419

In [30]:
query_5_5=f'''
    SELECT
        e.name AS employers,
        COUNT(i.industry_id) AS industry
    FROM public.employers e
        JOIN public.employers_industries i ON e.id = i.employer_id
    GROUP BY 1
    HAVING COUNT(i.industry_id) = 4   
    ORDER BY 1
    LIMIT 5'''

In [31]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employers  industry
0  101 Интернет         4
1      21vek.by         4
2          2ГИС         4
3            2К         4
4   4 пикселя +         4

In [32]:
query_5_6=f'''
    SELECT
        COUNT(e.name)
    FROM public.employers e
        JOIN public.employers_industries ei ON e.id = ei.employer_id
        JOIN public.industries i ON ei.industry_id = i.id
    WHERE i.name = 'Разработка программного обеспечения'   '''

In [33]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

In [34]:
query_5_7 = f'''
    SELECT
        a.name AS city_mill_yandex,
        count(v.area_id) AS cnt_vacan
        
    FROM
        employers e
     	JOIN vacancies v ON e.id = v.employer_id
        JOIN areas a ON a.id = v.area_id
     	      
    WHERE e.name = 'Яндекс'
        AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск',
        'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск',
        'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск',
        'Краснодар', 'Воронеж',  'Пермь', 'Волгоград')
    
    GROUP BY 1 
    
    UNION ALL

    SELECT 
        'Total',
        count(*)
        
    FROM 
        vacancies v
 	    JOIN employers e on e.id = v.employer_id
 	    JOIN areas a on a.id = v.area_id
      
    WHERE
        e.name = 'Яндекс'
        AND a.name IN ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
                 	'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск',
                 	'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар',
                    'Воронеж',  'Пермь', 'Волгоград') 
    ORDER BY 2'''

In [35]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city_mill_yandex  cnt_vacan
0              Омск         21
1         Челябинск         22
2        Красноярск         23
3         Волгоград         24
4             Пермь         25
5            Казань         25
6    Ростов-на-Дону         25
7               Уфа         26
8            Самара         26
9         Краснодар         30
10          Воронеж         32
11      Новосибирск         35
12  Нижний Новгород         36
13     Екатеринбург         39
14  Санкт-Петербург         42
15           Москва         54
16            Total        485

In [36]:
query_6_1=f'''
    SELECT
        COUNT(name) AS vacancy
    FROM public.vacancies
    WHERE name ILIKE'%data%' OR name ILIKE'%данн%'
    '''

In [37]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vacancy
0     1771

In [38]:
query_6_2=f'''
    SELECT 
        COUNT(name) AS vacancy
    FROM public.vacancies
    WHERE
        (name ILIKE '%data scientist%'
        or name ILIKE  '%data science%'
        or name ILIKE  '%исследователь данных%'
        or (name LIKE  '%ML%' and name NOT ILIKE '%HTML%')
        or name ILIKE  '%machine learning%'
        or name ILIKE  '%машинн%обучен%')
        
        AND (name ILIKE '%junior%'
        OR experience ILIKE '%Нет опыта%'
        OR employment ILIKE '%Стажировка%')
'''

In [39]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


vacancy
0       51

In [40]:
query_6_3 = f'''
    SELECT   COUNT(name) AS vacancy
        
    FROM  public.vacancies
     	      
    WHERE 
        (name ILIKE '%data scientist%'
        or name ILIKE  '%data science%'
        or name ILIKE  '%исследователь данных%'
        or (name LIKE  '%ML%' and name NOT ILIKE '%HTML%')
        or name ILIKE  '%machine learning%'
        or name ILIKE  '%машинн%обучен%')
        
        AND (key_skills ILIKE '%SQL%'
        OR key_skills ILIKE '%postgres%')
'''

In [41]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


vacancy
0      201

In [42]:
query_6_4 = f'''
    SELECT   COUNT(name) AS vacancy
        
    FROM  public.vacancies
     	      
    WHERE 
        (name ILIKE '%data scientist%'
        or name ILIKE  '%data science%'
        or name ILIKE  '%исследователь данных%'
        or (name LIKE  '%ML%' and name NOT ILIKE '%HTML%')
        or name ILIKE  '%machine learning%'
        or name ILIKE  '%машинн%обучен%')
        
        AND (key_skills ILIKE '%Python%')
'''

In [43]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


vacancy
0      351

In [44]:
query_6_5 = f'''
    SELECT  
        ROUND(AVG(length(key_skills)
        -length(replace(key_skills, CHR(9), ''))+1), 2) AS key_skills_in_vacanc
        
    FROM  vacancies
     	      
    WHERE 
        (name ILIKE '%data scientist%'
        or name ILIKE  '%data science%'
        or name ILIKE  '%исследователь данных%'
        or (name LIKE  '%ML%' and name NOT ILIKE '%HTML%')
        or name ILIKE  '%machine learning%'
        or name ILIKE  '%машинн%обучен%')
'''

In [45]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


key_skills_in_vacanc
0                  6.41

In [46]:
query_6_6 = f'''
    SELECT  
        DISTINCT experience,
        ROUND(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) AS mean_salary      
      
    FROM  vacancies
     	      
    WHERE 
        (name ILIKE '%data scientist%'
        or name ILIKE  '%data science%'
        or name ILIKE  '%исследователь данных%'
        or (name LIKE  '%ML%' and name NOT ILIKE '%HTML%')
        or name ILIKE  '%machine learning%'
        or name ILIKE  '%машинн%обучен%')
        AND (experience ILIKE '%от 3 до 6 лет%')
    
    GROUP BY 1
'''

In [47]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  mean_salary
0  От 3 до 6 лет     243115.0

Общий вывод по проекту

Рынок труда в различных городах почти одинаковый, в частности в городах-миллионниках.

Москва является лидером по общему числу работодателей

IT-сфера является самой распространенной среди вакансий, одни профессии связанные с данными состовляют 1771 вакансию

Компания Яндекс является самым крупным работодателем в числе городов-миллиоников

In [48]:
query_7 = f'''
    SELECT  
        name AS vacancies_IT,
         ROUND(AVG(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) AS mean_salary     
      
    FROM  vacancies
     	      
    WHERE 
        name LIKE  '%IT%'
        
    GROUP BY 1
    ORDER BY 2 DESC nulls last
    LIMIT 5
    
'''

In [49]:
df = pd.read_sql_query(query_7, connection)
df

C:\Users\nikon\AppData\Local\Temp\ipykernel_7468\2767996417.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7, connection)


vacancies_it  mean_salary
0                 Директор IT (HoReCa, Retail)     550000.0
1  Директор по информационным технологиям (IT)     320000.0
2                                  IT Директор     300000.0
3              IT специалист (PHP-разработчик)     300000.0
4               Руководитель IT инфраструктуры     250000.0

Выше предствлены самые высокооплачиваемые вакансии IT-сферы